### PR TITLE
CNI Ipv6 Dualstack Support

### PR DESCRIPTION
**Pull request by tamilmani1989**
**What this PR does / why we need it**:
This PR adds ipv6 datapath support for containers. The ipv6 ips are non-vnet ips and connection across VM works only with Azure udr solution. The containers will get ipv6 address inaddtion to vnet ipv4 address. containers within VM can communicate directly via ipv6. For Container communication across VMs, the src ip has to snat with vnet ipv6 CA.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

File: ebtables/ebtables.go [hunk 1]
```diff
@@ -133,6 +153,23 @@ func GetEbtableRules(tableName, chainName string) ([]string, error) {
 	return rules, nil
 }
 
+func SetBrouteAcceptCidr(ipNet net.IPNet, action string) error {
```
Comment (matmerr) [inline ebtables/ebtables.go:156]: Add comment here

File: cni/network/network.go [hunk 2]
```diff
@@ -750,6 +833,24 @@ func (plugin *netPlugin) Delete(args *cniSkel.CmdArgs) error {
 		return err
 	}
 
+	defer func() {
+		msg = fmt.Sprintf("CNI DEL succeeded : Released ip %+v podname %v namespace %v", nwCfg.Ipam.Address, k8sPodName, k8sNamespace)
+		if err != nil {
+			msg = fmt.Sprintf("CNI DEL failed : ip %+v podname %v namespace %v", nwCfg.Ipam.Address, k8sPodName, k8sNamespace)
+		}
+
+		plugin.setCNIReportDetails(nwCfg, CNI_DEL, msg)
```
Comment (matmerr) [inline cni/network/network.go:842]: This change may not be required

File: network/network_linux.go [hunk 3]
```diff
@@ -450,6 +466,34 @@ func (nm *networkManager) disconnectExternalInterface(extIf *externalInterface,
...
+				return epcommon.AssignIPToInterface(nwInfo.BridgeName, ipAddr)
+			}
+		}
+	}
+
+	return nil
+}
+
+func addIpv6SnatRule(extIf *externalInterface) error {
+	log.Printf("[net] Adding ipv6 snat rule")
```
Comment (matmerr) [inline network/network_linux.go:487]: add conditional based on mode

File: network/network_linux.go [hunk 4]
```diff
@@ -420,16 +421,22 @@ func (nm *networkManager) connectExternalInterface(extIf *externalInterface, nwI
...
 			return err
 		}
 
-		if err = addIpv6SnatRule(extIf, nwInfo.IPV6Mode); err != nil {
+		if err = addIpv6SnatRule(extIf, nwInfo); err != nil {
 			log.Errorf("[net] Adding IPv6 Snat Rule failed:%v", err)
 			return err
 		}
+
+		// unmark packet if set by kube-proxy
```
Comment (matmerr) [inline network/network_linux.go:435]: can you add a more in depth explanation why this mangle rule is needed?

File: network/epcommon/endpoint_common.go [hunk 5]
```diff
@@ -28,8 +28,9 @@ RFC for Link Local Addresses: https://tools.ietf.org/html/rfc3927
 */
 
 const (
-	enableIPForwardCmd = "sysctl -w net.ipv4.ip_forward=1"
-	enableIPV6Cmd      = "sysctl -w net.ipv6.conf.all.disable_ipv6=%d"
+	enableIPForwardCmd   = "sysctl -w net.ipv4.ip_forward=1"
+	enableIPV6Cmd        = "sysctl -w net.ipv6.conf.all.disable_ipv6=%d"
```
Comment (matmerr) [inline network/epcommon/endpoint_common.go:32]: Potentially rename to toggleIPV6, given that cmd usage could be to enable or disable

File: netlink/link.go [hunk 6]
```diff
@@ -391,7 +391,7 @@ func SetLinkHairpin(bridgeName string, on bool) error {
 }
 
 // AddOrRemoveStaticArp sets/removes static arp entry based on mode
-func AddOrRemoveStaticArp(mode int, name string, ipaddr net.IP, mac net.HardwareAddr) error {
+func AddOrRemoveStaticArp(mode int, name string, ipaddr net.IP, mac net.HardwareAddr, isProxy bool) error {
```
Comment (matmerr) [inline netlink/link.go:394]: can you give a quick summary what the NTF flag is for?

File: network/endpoint.go [hunk 7]
```diff
@@ -82,6 +85,7 @@ type RouteInfo struct {
 	Protocol int
 	DevName  string
 	Scope    int
+	Priority int
```
Comment (matmerr) [inline network/endpoint.go:88]: can we keep this as metric and instead, and add a comment to say that it corresponds to route metric?

File: netlink/link.go (hunk 6 above)
Comment (tamilmani1989) [reply netlink/link.go:394]: sure..its for neighbor proxy..not using right now

File: network/endpoint.go (hunk 7 above)
Comment (tamilmani1989) [reply network/endpoint.go:88]: golang netlink package names it as priority..keeping it consistent with that

**Pull request merged**